### PR TITLE
A draft of ip range visitor to minimize requests to geo range.

### PR DESCRIPTION
**Pull request by 3asm**

### CODE REVIEW

File: agent/ip2geo_agent.py [hunk 1]
```diff
@@ -20,41 +19,54 @@
...
 
-GEO_LOCATION_API_ENDPOINT =  'http://ip-api.com/json/'
-STORAGE_NAME = 'agent_ipgeo_storage'
+
+class Error(Exception):
+    """Base Error"""
+
+
+class IPGeoError(Error):
+    """Error while finding ips geolocation"""
```
Comment (3asm) [inline agent/ip2geo_agent.py:30]:
```suggestion
    """Error getting the IP geolocation"""
```

File: agent/ip2geo_agent.py [hunk 2]
```diff
@@ -20,41 +19,54 @@
...
-                out_selector = f'{message.selector}.geolocation'
-                self.emit(selector=out_selector, data=geolocation_details)
-        else:
-            logger.info('%s has already been processed. skipping for now.', ip_address)
-
+        ip = message.data['host']
+        out_selector = f'{message.selector}.geolocation'
+
+        mask = message.data.get('mask', '32')
+        network = ipaddress.ip_network(f"""{ip}/{mask}""")
```
Comment (3asm) [inline agent/ip2geo_agent.py:54]:
```suggestion
        network = ipaddress.ip_network(f'{ip}/{mask}')
```

File: agent/ip2geo_agent.py [hunk 3]
```diff
@@ -20,41 +19,54 @@
...
+
+        # classify ip range based on geolocation
+        for result in ip_range_visitor.dichotomy_ip_network_visit(
+                network, ip_range_visitor.is_first_last_ip_same_geolocation):
+
+            geolocation_details = result[0]
+            network = result[2]
+            for ip in network:
+                # check if ip not tested before
+                if self.set_add(STORAGE_NAME, ip) is True:
```
Comment (3asm) [inline agent/ip2geo_agent.py:64]: Please use the new mixin set_ip_range.

File: agent/request_sender.py [hunk 4]
```diff
@@ -29,7 +29,7 @@ def make_request(method: str, path: str, data: Optional[Dict[str, str]] = None):
         AuthenticationError if request is not successful.
     """
     logger.info('request %s %s %s', method, path, data)
-    response = requests.request(method, path, data=json.dumps(data))
+    response = requests.request(method, path, data=json.dumps(data), timeout=None)
```
Comment (3asm) [inline agent/request_sender.py:32]:
```suggestion
    response = requests.request(method, path, data=json.dumps(data), timeout=10)
```

File: tests/utils/test_ip_range_visitor.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
+"""Unittests for the Ip range visitor."""
```
Comment (3asm) [inline tests/utils/test_ip_range_visitor.py:1]:
```suggestion
"""Unittests for the IP range visitor."""
```

File: tests/utils/test_ip_range_visitor.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+import ipaddress
+
+from agent.utils import ip_range_visitor
+
+
+def testVistor_withMatchingIPAndMaskRecieved_returnsLocations():
+    results = []
+    for result in ip_range_visitor.dichotomy_ip_network_visit(ipaddress.ip_network('8.8.8.0/22'),
+                    ip_range_visitor.is_first_last_ip_same_geolocation):
+        print(result)
```
Comment (3asm) [inline tests/utils/test_ip_range_visitor.py:12]:
```suggestion
```

File: tests/utils/test_ip_range_visitor.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        results.append(result)
+
+    assert all(len(result) == 3 and result[0] != result[1] for result in results)
```
Comment (3asm) [inline tests/utils/test_ip_range_visitor.py:15]: This is not really checking the location output. You need to check the value returned.

File: tests/utils/test_ip_range_visitor.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+
+def testVistor_withMatchingIPAndMaskNotRecieved_returnsLocation():
```
Comment (3asm) [inline tests/utils/test_ip_range_visitor.py:18]: The name of the unit test do not match test case.

**Review by najibraihan — APPROVED** [review]

File: agent/ip2geo_agent.py [hunk 9]
```diff
@@ -42,19 +41,26 @@ def process(self, message: m.Message) -> None:
...
+        ip = message.data['host']
+        mask = message.data.get('mask', '32')
+        network = ipaddress.ip_network(f'{ip}/{mask}')
+
+        # classify ip range based on geolocation
+        for result in self.ip_range_visitor.dichotomy_ip_network_visit(
+                network, self.ip_range_visitor.is_first_last_ip_same_geolocation):
+
+            geolocation_details = result[0]
+            network = result[2]
```
Comment (amine3) [inline agent/ip2geo_agent.py:55]: Can you change this name to avoid confusion with network defined in line 48

**Review by amine3 — APPROVED** [review]

**Pull request merged**